clang/lib/Sema/SemaType.cpp

```diff
@@ -8938,6 +8957,11 @@ static void processTypeAttrs(TypeProcessingState &state, QualType &type,
       attr.setUsedAsTypeAttr();
       break;
     }
+    case ParsedAttr::AT_RequiresCapability: {
+      HandleRequiresCapabilityAttr(state, type, attr);
+      attr.setUsedAsTypeAttr();
+      break;
+    }
```

Comment:
Okay, maybe we do want to do this change separately. I think I was wrong about modeling after this; I think we want it to be done like `handleFunctionTypeAttr()` and `FUNCTION_TYPE_ATTRS_CASELIST` so that it becomes part of the `FunctionType` itself. That's what would get us the diagnostics on type mismatches (we look through `AttributedType` as sugar: https://github.com/llvm/llvm-project/blob/28b7e281d4eaea0d5d56b1a4cf7a550be746a007/clang/include/clang/AST/Type.h#L4999)

Reply:
Hmmmm, my intuition is that if we don't do this work now, we'll forever be stuck with the current behavior where we don't diagnose pointer mismatches because someone will rely on that continuing to work. So maybe we do want to do this work now...

CC @erichkeane @aaronpuchert for opinions on whether we think this needs to diagnose mismatches in the initial offering of it as a type attribute.

Reply:
I think I agree Aaron, if we don't diagnose now, it'll be pretty ugly to do it later.

Reply:
Meh, it seems like adding a pointer to `FunctionType::ExtInfo` would blow it up a lot, and just a bit isn't enough to save the information.

Reply:
There were a few changes there recently... i wonder if it is time to just accept that we're going all-in on trailing-storage for this type.  We're now doing it for the ARM SVE types, plus the 'lesser used' structure, so perhaps the 'capability pointer' can end up there too.  WDYT?

Reply:
@AaronBallman Does this sound like the way to go to you as well?

Reply:
No, we don't want to modify `FunctionType::ExtInfo`, I think what @erichkeane was talking about is that we added trailing objects to `FunctionProtoType` that we can squirrel lesser-used data into. That might be reasonable here because I cannot imagine anyone using thread safety attributes with K&R C interfaces, but that would be the downside to putting the data on `FunctionProtoType`.

Reply:
It would kind of stink to have to do it in 2 places, but it could go in trailing-storage/etc on `FunctionProtoType` and `FunctionNoProtoType`, then just disambiguate it with a function.  So it doesn't need to end up in `FunctionType::ExtInfo` or `FunctionType` itself.